File: java/src/main/java/org/example/sudoku_validator/SudokuValidator.java
package org.example.sudoku_validator;

import java.util.HashSet;
import java.util.Set;

public class SudokuValidator {
    public static boolean isValidSudoku(char[][] board) {
        // Check rows
        for (int row = 0; row < 9; row++) {
            if (isValid(board[row])) {
                return false;
            }
        }

        // Check columns
        for (int col = 0; col < 9; col++) {
            char[] column = new char[9];
            for (int row = 0; row < 9; row++) {
                column[row] = board[row][col];
            }
            if (isValid(column)) {
                return false;
            }
        }

        // Check sub-boxes
        for (int i = 0; i < 3; i++) {
            for (int j = 0; j < 3; j++) {
                char[] subBox = new char[9];
                int index = 0;
                for (int row = i * 3; row < (i + 1) * 3; row++) {
                    for (int col = j * 3; col < (j + 1) * 3; col++) {
                        subBox[index++] = board[row][col];
                    }
                }
                if (isValid(subBox)) {
                    return false;
                }
            }
        }

        return true;
    }

    private static boolean isValid(char[] nums) {
        Set<Character> digits = new HashSet<>();
        for (char num : nums) {
            if (num != '.') {
                if (digits.contains(num)) {
                    return true;
                }
                digits.add(num);
            }
        }
        return false;
    }
}

File: python/sudoku_validator/SudokuValidator.py
class Solution:
    def isValidSudoku(self, board):
        # Check rows
        for row in board:
            if not self.is_valid(row):
                return False

        # Check columns
        for col in range(9):
            column = [board[i][col] for i in range(9)]
            if not self.is_valid(column):
                return False

        # Check sub-boxes
        for i in range(0, 9, 3):
            for j in range(0, 9, 3):
                sub_box = [board[x][y] for x in range(i, i + 3) for y in range(j, j + 3)]
                if not self.is_valid(sub_box):
                    return False

        return True
    def is_valid(self, nums):
        seen = set()
        for num in nums:
            if num != ".":
                if num in seen:
                    return False
                seen.add(num)
        return True
